### PR TITLE
Enhanced : Test for pkg/gofr/websocket module

### PR DESCRIPTION
**Pull request by thzgajendra**
# WebSocket Package: 100% Test Coverage & Linter Fixes

## 📋 Overview

This PR implements comprehensive test coverage for the `pkg/gofr/websocket` package, achieving **100% test coverage** while fixing all linter issues. The changes include extensive test cases covering all functionality, edge cases, performance scenarios, and proper error handling.

## 🎯 Goals Achieved

- ✅ **100% Test Coverage** - All functions and methods in the websocket package are fully tested
- ✅ **Zero Linter Issues** - All golangci-lint issues resolved without disabling any linters
- ✅ **Comprehensive Test Suite** - Tests cover normal operations, edge cases, error conditions, and performance
- ✅ **Code Quality** - Proper error handling, thread safety, and clean code practices

## 📊 Test Coverage Results

```bash
PASS
coverage: 100.0% of statements
ok      gofr.dev/pkg/gofr/websocket     2.143s
```

### Coverage Breakdown:
- **websocket.go**: 100% coverage across all 20 functions
- **options.go**: 100% coverage across all 7 option functions  
- **mock_interfaces.go**: 100% coverage across all 4 mock functions

### 1. **Consolidated Test File**
- **File**: `pkg/gofr/websocket/websocket_test.go`
- **Lines**: 986 lines of comprehensive test coverage
- **Approach**: Consolidated all test cases into a single, well-organized test file

### 2. **Core Functionality Tests**
- ✅ **WSUpgrader Tests**: Creation, options, upgrade operations
- ✅ **Connection Tests**: Bind, WriteMessage, ReadMessage, deadlines

### 3. **Edge Case Coverage**
- ✅ **Error Handling**: Invalid JSON, closed connections, nil pointers
- ✅ **Concurrent Operations**: Thread-safe operations with proper synchronization
- ✅ **Performance Tests**: Large message handling, concurrent writes

### **Thread Safety**
- Concurrent read/write operations
- Thread-safe connection management
- Proper mutex usage verification

### **Error Handling**
- Invalid JSON parsing
- Closed connection handling
- Network error scenarios
- Upgrade failure handling

### **Performance**
- Large message throughput
- Concurrent connection scaling
- Memory usage optimization

### **WebSocket Protocol**
- Text message handling
- Connection lifecycle management
- Deadline enforcement
- Proper cleanup

### **Run All Tests**
```bash
cd pkg/gofr/websocket
go test -v
```

### **Run with Coverage**
```bash
go test -cover -coverprofile=coverage.out
go tool cover -func=coverage.out
```

### **Run Performance Tests**
```bash
go test -v -short=false
```

### **Run Linter**
```bash
golangci-lint run .
```

## 📈 Metrics

- **Total Test Functions**: 25+
- **Test Coverage**: 100%
- **Linter Issues**: 0
- **Test Execution Time**: ~2.1 seconds
- **Lines of Test Code**: 986

### **Error Handling**
- Proper error checking in HTTP handlers
- Graceful error recovery in tests
- Clear error messages and assertions

### **Code Organization**
- Logical test grouping by functionality
- Clear test naming conventions
- Comprehensive test documentation

### **Performance**
- Efficient test execution
- Minimal resource usage
- Proper cleanup and teardown

## ✅ Checklist

- [x] **100% Test Coverage Achieved**
- [x] **All Linter Issues Fixed**
- [x] **Core Functionality Tested**
- [x] **Edge Cases Covered**
- [x] **Error Handling Verified**
- [x] **Thread Safety Tested**
- [x] **Performance Tests Included**
- [x] **Code Quality Maintained**
- [x] **Documentation Updated**
- [x] **No Breaking Changes**

---

**Ready for Review** ✅

Screenshot 2025-09-22 at 6 30 22 PM

### CODE REVIEW

**Review by coolwednesday — CHANGES_REQUESTED** [review]
Here are some areas for improvement in the tests:

1. **Error Handling Coverage**:
   - Some tests, like `TestConnection_Bind_ErrorPaths`, could include more diverse scenarios for error handling, such as network interruptions or unexpected server responses.

4. **Mock Usage**:
   - Some unrelated tests use real WebSocket connections unnecessarily. Mocking could improve test speed and isolation, especially in `TestConnection_Bind` and `TestConnection_ReadMessage`. Please note not necessarily for all tests.

5. **Test Duplication**:
   - There is some duplication in setup code for WebSocket servers. Extracting this into reusable helper functions could improve maintainability. Add T.Helper for them.

6. **Unimplemented Methods**:
   - Tests for unimplemented methods like `Param` and `PathParam` are redundant unless these methods are expected to be implemented in the future.

9. **Test Coverage for Options**:
   - Ensure all combinations of options in `NewWSUpgrader` are tested, especially for conflicting or invalid configurations. - Conflicting options should give error. (this will require real websocket connection)
   
   
   Other than this, @thzgajendra, Good work !

**Comment by thzgajendra** [comment]
## Test Improvements in WebSocket Package

This PR addresses multiple gaps and improvements in the WebSocket test suite:

- [ ] Error Handling Coverage
- Expanded `TestConnection_Bind_ErrorPaths` to cover more diverse error handling cases, including:
  - Network interruptions
  - Unexpected server responses

- [ ] Mock Usage
- Replaced some real WebSocket connections with **mocks** where appropriate.  
  - This improves **test speed** and **isolation**.  
  - Example: `TestConnection_Bind` and `TestConnection_ReadMessage`.  
  - (Note: not applied to all tests — only where relevant.)

- [ ] Test Duplication
- Reduced duplication of setup code for WebSocket servers.  
- Extracted common setup logic into **reusable helper functions**.  
- Added `t.Helper()` annotations for better test stack traces.

- [ ] Unimplemented Methods
- Removed redundant tests for **unimplemented methods** (`Param`, `PathParam`).  
- Tests will be added later **only if these methods are implemented**.

- [ ] Test Coverage for Options
- Ensured all combinations of **options in `NewWSUpgrader`** are tested:
  - Valid configurations
  - Conflicting/invalid configurations → expected to return error  
  - (Note: conflict validation requires a real WebSocket connection, so those tests are kept with real connections.)


---

Overall, this PR improves **coverage, maintainability, and reliability** of the WebSocket test suite while removing unnecessary or redundant tests.

File: pkg/gofr/websocket/mock_interfaces.go [hunk 1]
```diff
@@ -1,9 +1,7 @@
 // Code generated by MockGen. DO NOT EDIT.
-// Source: interfaces.go
```
Comment (Umang01-hash) [inline pkg/gofr/websocket/mock_interfaces.go:2]: Why they are being removed??

File: pkg/gofr/websocket/websocket_test.go [hunk 2]
```diff
@@ -11,221 +14,1706 @@ import (
...
 
-				wsConn := &Connection{Conn: conn}
-
-				var data any
-				switch tt.expectedData.(type) {
-				case string:
-					data = new(string)
-				default:
-					data = &map[string]any{}
+				if tt.expectError {
```
Comment (Umang01-hash) [inline pkg/gofr/websocket/websocket_test.go]: Test methods should not contains if else statements like this. If required we can break the method into two methods one for success case and other for error.

File: pkg/gofr/websocket/websocket_test.go [hunk 3]
```diff
@@ -11,221 +14,1706 @@ import (
...
-				wg.Add(1)
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			upgrader := NewWSUpgrader(tt.options...)
 
-				go func() {
-					defer wg.Done()
+			server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				conn, err := upgrader.Upgrade(w, r, nil)
+				if tt.expectError {
```
Comment (Umang01-hash) [inline pkg/gofr/websocket/websocket_test.go]: Test methods should not contains if else statements like this. If required we can break the method into two methods one for success case and other for error.

File: pkg/gofr/websocket/websocket_test.go [hunk 4]
```diff
@@ -11,221 +14,1706 @@ import (
...
+			}))
+			defer server.Close()
 
-		wg.Wait()
-	}))
+			url := "ws" + server.URL[len("http"):] + "/ws"
+			dialer := websocket.DefaultDialer
+			conn, resp, err := dialer.Dial(url, nil)
+
+			if tt.expectError {
```
Comment (Umang01-hash) [inline pkg/gofr/websocket/websocket_test.go]: Remove this if else also please

File: pkg/gofr/websocket/websocket_test.go [hunk 5]
```diff
@@ -11,221 +14,1706 @@ import (
...
-	manager.AddWebsocketConnection("conn3", &Connection{Conn: &websocket.Conn{}})
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			upgrader := NewWSUpgrader(tt.options...)
 
-	// Get the list of connections
-	connections := manager.ListConnections()
+			server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				conn, err := upgrader.Upgrade(w, r, nil)
+				if tt.expectError {
```
Comment (Umang01-hash) [inline pkg/gofr/websocket/websocket_test.go]: Here also we can break it into two method with and without error but no if else like this.

File: pkg/gofr/websocket/websocket_test.go [hunk 6]
```diff
@@ -11,221 +14,1706 @@ import (
...
 
-func TestManager_GetConnectionByServiceName(t *testing.T) {
-	manager := New()
+			url := "ws" + server.URL[len("http"):] + "/ws"
+			dialer := websocket.DefaultDialer
+			conn, resp, err := dialer.Dial(url, nil)
 
-	mockConn := &Connection{Conn: &websocket.Conn{}}
-	manager.AddWebsocketConnection("testService", mockConn)
+			if tt.expectError {
```
Comment (Umang01-hash) [inline pkg/gofr/websocket/websocket_test.go]: Here also

File: pkg/gofr/websocket/websocket_test.go [hunk 7]
```diff
@@ -11,221 +14,1706 @@ import (
...
+					t.Errorf("Failed to upgrade connection: %v", err)
+					return
+				}
+
+				defer conn.Close()
+
+				wsConn := &Connection{Conn: conn}
+				err = wsConn.Bind(tt.targetType)
+
+				if tt.expectError {
```
Comment (Umang01-hash) [inline pkg/gofr/websocket/websocket_test.go]: Please refactor this too.

**Comment by thzgajendra** [comment]
Sure , I'll Update these recommended changes.

**Review by Umang01-hash — APPROVED** [review]

**Review by Umang01-hash — APPROVED** [review]

**Comment by thzgajendra** [comment]
@coolwednesday updated.  I've refactored all test helpers and test cases to use GoFr's `*Connection` type instead of gorilla's `*websocket.Conn`.
